verilog/tools/kythe/indexing_facts_tree_extractor_test.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "indexing_facts_tree_extractor.h"
+#include "common/text/concrete_syntax_tree.h"
+#include "gtest/gtest.h"
+#include "verilog/analysis/verilog_analyzer.h"
+
+#undef EXPECT_OK
+#define EXPECT_OK(value) EXPECT_TRUE((value).ok())
+
+namespace verilog {
+namespace kythe {
+namespace {
+
+typedef IndexingFactNode T;
+
+bool FactsTreeEqual(const IndexingNodeData& lhs, const IndexingNodeData& rhs) {
+  auto left = lhs, right = rhs;
+  bool equal = left.GetIndexingFactType() == right.GetIndexingFactType();
+
+  const auto &lAnchors = left.Anchors(), rAnchors = right.Anchors();
+  equal &= lAnchors.size() == rAnchors.size();
+  for (size_t i = 0; i < lAnchors.size(); i++) {
+    equal &= lAnchors[i].StartLocation() == rAnchors[i].StartLocation() &&
+             lAnchors[i].EndLocation() == rAnchors[i].EndLocation() &&
+             lAnchors[i].Value() == rAnchors[i].Value();
+  }
+
+  return equal;
+}
+
+TEST(EmptyCSTTest, FactsTreeExtractor) {
+  std::string code_text = "";
+  VerilogAnalyzer analyzer(code_text, "");
+  EXPECT_OK(analyzer.Analyze());
+  const auto& root = analyzer.Data().SyntaxTree();
+  const absl::string_view file_name = "verilog.v";
+
+  const auto expected =
+      T({{Anchor(file_name, 0, code_text.size())}, IndexingFactType ::kFile});
+
+  const auto& facts_tree = BuildIndexingFactsTree(
+      ABSL_DIE_IF_NULL(root), analyzer.Data().Contents(), file_name);
+
+  const auto result_pair = DeepEqual(facts_tree, expected, FactsTreeEqual);
+  EXPECT_EQ(result_pair.left, nullptr) << *result_pair.left;
+  EXPECT_EQ(result_pair.right, nullptr) << *result_pair.right;
+}
+
+TEST(EmptyModuleTest, FactsTreeExtractor) {
+  std::string code_text = "module foo; endmodule: foo";
+  VerilogAnalyzer analyzer(code_text, "");
+  EXPECT_OK(analyzer.Analyze());
+  const auto& root = analyzer.Data().SyntaxTree();
+  const absl::string_view file_name = "verilog.v";
+
+  const auto expected =
+      T({{Anchor(file_name, 0, code_text.size())}, IndexingFactType ::kFile},
+        T({{Anchor(absl::string_view("foo"), 7, 10),
+            Anchor(absl::string_view("foo"), 23, 26)},
+           IndexingFactType::kModule}));
```

Comment:
This is good for now.

Suggested study, that may make future test maintenance more productive:

https://github.com/google/verible/blob/b1027eedd092c8cf6162f4223b2d9c861e23b250/verilog/CST/declaration_test.cc#L80
https://github.com/google/verible/blob/b1027eedd092c8cf6162f4223b2d9c861e23b250/verilog/CST/declaration_test.cc#L116

This is an example of embedding expected subranges into the test case itself, which has supporting functions to automatically calculates offsets for you when piecing together the strings.

This only works for **non-overlapping** expected regions, and I haven't gotten around to making it work with nested overlapping ranges yet.  (big TODO for me).

Wouldn't it be nice to write something like:

```c++
{kModuleDeclaration, {"module ", {kModuleName, "foo"}, "; endmodule: ", {kModuleName, "foo"}, "\n"}}
```

and have it build the expected tree for you?

Reply:
Could u provide some more details?

Reply:
The concept here is to be able to write an annotated tree of string snippets and have that generate both the test input and test expected results.  In this case, that would automate those byte offset numbers for you.  Don't worry about this right now.